code/modules/GOAP/GOAPAction.dm

```diff
@@ -0,0 +1,80 @@
+// macro to check if the agent can path to atom X in the action returns TRUE if its possible and FALSE if its not, this won't work with turfs
+#define PATH_CHECK(action, X) !(X.loc in action.bturfs)
+
+/datum/goap_action
+	var/name = "generic goap action" //for bookkeeping/debug
+	var/list/preconditions //Our requirements from the World State
+	var/list/effects //Our effects on the World State
+	var/list/bturfs = list() // turfs where pathing has failed for this action
```

Comment:
lazylist this?

Reply:
It would be the same and I don't think you can use that macro here due to it using an if statement

Reply:
lazylists work fine with if statements, LAZYADD() LAZYLEN() LAZYREMOVE() etc

Reply:
benefit is potentially less empty lists taking up memory